### DISCUSSION TITLE
Guide for how language communities can self-serve language support?

### DISCUSSION THREAD

**Discussion by elliette** [community/community#10934] [Other Feature Feedback, Questions, & Ideas] [answered]
Hi! I've been enjoying using Github Code Search so far. 

I'm part of the Dart language team (https://github.com/dart-lang) and we would love to add Dart support for Github Code Search. 

* I see Code Search is currently using the open source Tree-Sitter parser tool: https://cs.github.com/about/faq#languages
* I also see that Github code navigation outside of Code Search uses Tree-sitter in combination with stack-graphs for precise code navigation (it looks like only Python is supported at the moment): https://github.blog/2021-12-09-precise-code-navigation-python-code-navigation-pull-requests 

In the [announcement about precise code navigation](https://github.blog/2021-12-09-precise-code-navigation-python-code-navigation-pull-requests/) it's stated:
> "We will also be publishing information on how language communities can self-serve stack graph support for their languages, should they wish to."

I was wondering if that guide has been published yet? Or if not, I wanted to verify that if we add Tree-sitter support for Dart, the Github team would be able to add symbol extraction in Code Search and in the regular Github code UI? (Since stack-graphs builds off of the Tree-sitter parsers this seems like a reasonable first step). 

Thank you!

**Comment by colinwm** [comment, marked as answer]
For code search, if you implement a tree-sitter parser for Dart, we can absolutely enable things like fuzzy jump-to-definition and better ranking. That would be awesome! We can do it pretty quickly, just let us know when you have a parser.

Getting "precise" stack graph support is a bit more complex, as it's more than just a parser. But we want to enable third parties to help us with that too, I'll let @dcreager answer that one.

**Reply by dcreager** [reply]
Hi there!  Yes as @colinwm mentioned, the first step is to get a tree-sitter parser that you're happy with, including syntax highlighting and symbol tagging queries.  That will be enough for us to support Dart in Code Search, Fuzzy Code Nav, and Syntax Highlighting.  (Dart is already a syntax-highlighted, but uses an old highlighting engine based on regular expressions.  As part of adding support for Code Nav and friends, we typically migrate the language over to our new parser-based highlighting engine.)

For stack graphs, we still haven't published a proper guide yet, but the [tree-sitter-graph](https://github.com/tree-sitter/tree-sitter-graph) and [stack-graphs](https://github.com/github/stack-graphs) repos are the best place to follow our progress.  The tl;dr is that we've created a new "graph construction DSL" for tree-sitter, which lets you create the necessary stack graph structure from the parsed syntax of the language.  (If you haven't seen it yet, I go into some initial detail about that process towards the end of my [Strange Loop talk](https://dcreager.net/talks/2021-strange-loop/) from last fall.)  Those two repos are where that logic will live — we are migrating an internal version of that DSL over to the externally visible tree-sitter repos.

Re timing, my hope is that if you all are able to make progress on the Fuzzy Code Nav side, by the time that is done we will hopefully be ready to work with some initial language communities to walk them through the process of writing stack graph rules.  So first steps are right in line with what @colinwm said — let's get a Dart grammar squared away with highlighting and tagging queries in it.

**Reply by elliette** [reply]
Great, thanks for all this great information! Will circle back around once we have a tree-sitter parser.
